core/container/externalbuilder/copy.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package externalbuilder
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/pkg/errors"
+)
+
+// MoveOrCopyDir attempts to copy src to dest by firstly trying to move, then copy upon failure.
+func MoveOrCopyDir(logger *flogging.FabricLogger, srcroot, destroot string) error {
+	mvErr := os.Rename(srcroot, destroot)
+	if mvErr == nil {
+		return nil
+	}
+
+	logger.Warnf("Failed to move %s to %s: %s, try copy instead", srcroot, destroot, mvErr)
+
+	info, err := os.Stat(srcroot)
+	if err != nil {
+		return errors.WithMessagef(err, "failed to stat")
+	}
+
+	if err = os.MkdirAll(destroot, info.Mode()); err != nil {
+		return errors.WithMessage(err, "failed to make dir")
+	}
+
+	cpErr := CopyDir(srcroot, destroot)
+	if cpErr == nil {
+		return nil
+	}
+
+	logger.Errorf("Failed to copy %s to %s: %s", srcroot, destroot, cpErr)
+
+	rmErr := os.RemoveAll(destroot)
+	if rmErr != nil {
+		logger.Errorf("Failed to clean targeting dir %s: %s", destroot, rmErr)
+	}
+
+	return errors.WithMessagef(cpErr, "failed to copy %s to %s", srcroot, destroot)
+}
+
+// CopyDir creates a copy of a dir
+func CopyDir(srcroot, destroot string) error {
+	return filepath.Walk(srcroot, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		srcsubpath, err := filepath.Rel(srcroot, path)
+		if err != nil {
+			return err
+		}
+		destpath := filepath.Join(destroot, srcsubpath)
+
+		if info.IsDir() { // its a dir, make corresponding dir in the dest
+			if err = os.MkdirAll(destpath, info.Mode()); err != nil {
+				return err
+			}
+			return nil
+		}
+
+		// its a file, copy to corresponding path in the dest
+		if err = copyFile(path, destpath); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+func copyFile(srcpath, destpath string) error {
```

Comment:
This doesn't seem to appropriately handle sym-links.  Maybe that's okay, but seems like a notable limitation.  But I'd welcome feedback.

https://stackoverflow.com/questions/51779243/copy-a-folder-in-go pointed to https://github.com/otiai10/copy/blob/master/copy.go which seemed instructive.

Reply:
yes, i was thinking of symlinks but not sure if that's necessary in our case. wdyt?

Reply:
i was thinking of symlinks but not sure if symlink is needed in our case, wdyt?

Reply:
@sykesm Do you think this is something we can punt on, or should try to support out of the gate?

Reply:
Discussed this with @sykesm and not handling symlinks should be fine for now.  If the link is relative and to a file, then we should get normal copy behavior.  There's no obvious reason to expect symlinks in the build output.